### DISCUSSION TITLE
Significant Slowdown in Simulation Time in DAG-OpenMC

### DISCUSSION THREAD

**Discussion by wrkendrick** [svalinn/DAGMC#953] [Q&A]
Hey all, I've used DAGMC for OpenMC neutronic simulations for while now, and I've always found that my DAG-OpenMC runs would be about an order of magnitude slower in particle simulation speed. I assumed it to be because of the complexity of representing geometries via facets, slowing down the ray tracing. Now, however, with the wonderful new way of using Trimesh as a basis for the geometry, I'm still seeing the same slowdown, even though all my common sense is telling me I should see a speed increase. I'm wondering if maybe the slowdown isn't caused by an inherent DAGMC property, but instead some mistake I'm making during installation? Would someone mind running the following two pincell input files (one with CSG geometry and one with DAGMC) and telling me what your change in particle simulation rate was? I've also attached the .h5m I use in the DAG-OpenMC case.

Any help is appreciated, thanks for your time!
[speed_comparison.zip](https://github.com/svalinn/DAGMC/files/14808997/speed_comparison.zip)

My outputs from the two are
DEFAULT:
Calculation Rate (inactive)       = 19863.3 particles/second
 Calculation Rate (active)         = 15876.6 particles/second
DAGMC TRI:
Calculation Rate (inactive)       = 1225.35 particles/second
 Calculation Rate (active)         = 1376.5 particles/second

**Comment by pshriwise** [comment]
Hi @wrkendrick. Thanks for giving the integrated DAGMC export a go! 

There are two things that come to mind you can try to improve DAGMC performance: 

1. Make sure to set the CMAKE_BUILD_TYPE variable to Release when building MOAB. The library can be extremely slow without the compiler optimization flags that setting provides. 

2. Try enabling the [double-down](https://github.com/pshriwise/double-down) interface (used a faster ray tracer than MOAB's). Once double-down is installed, you can enable the DOUBLE_DOWN option in the CMake command for DAGMC to do so.

**Reply by wrkendrick** [reply]
Thanks for the hints (and thank you @gonuke as well)! I'll give these changes a shot and get back to y'all with how it changes my simulation speed.

**Reply by ahnaf-tahmid-chowdhury** [reply]
@pshriwise, would enabling double-down for the DAGMC library boost the performance of PyNE as well, considering PyNE has a DAGMC module?

**Comment by gonuke** [comment]
I haven't tried running your problems, but it is true that DAGMC often appears to be slow when compared to CSG on the simplest geometries.  I think that even with @pshriwise's suggestions above, this will continue to be true.  DAGMC was build to support geometries that are either not possible (surfaces that can't be represented in CSG) or not practical (CAD-based models with too many surfaces/volumes to convert by hand) to represent by CSG. 

Of course, it is hard to compare performance for very complex models because they either can't be represented in CSG as a basis for comparison, or nobody wants to invest the effort to convert such a geometry.  There have been a few isolated comparisons with complex geometries that are converted using tools such as McCAD, but it has been some time since we repeated that.

In your case of a simple cylinder, I expect the CSG to be very fast compared to DAGMC, and wouldn't be surprised by a factor of 10.  The things suggested by @pshriwise may help by 2x-3x?
